### PR TITLE
Markduckworth/or queries pr 4

### PR DESCRIPTION
**Pull request by MarkDuckworth**
Hey there! So you want to contribute to a Firebase SDK? 
Before you file this pull request, please read these guidelines:

### Discussion

  * Read the contribution guidelines (CONTRIBUTING.md).
  * If this has been discussed in an issue, make sure to link to the issue here. 
    If not, go file an issue about this **before creating a pull request** to discuss.

### Testing

  * Make sure all existing tests in the repository pass after your change.
  * If you fixed a bug or added a feature, add a new test to cover your code.

### API Changes

  * At this time we cannot accept changes that affect the public API.  If you'd like to help 
    us make Firebase APIs better, please propose your change in an issue so that we 
    can discuss it together.

### CODE REVIEW

File: packages/firestore/test/unit/core/query.test.ts [hunk 1]
```diff
@@ -800,6 +803,71 @@ describe('Query', () => {
...
+    assertQueryMatches(query5, [doc3], [doc1, doc2, doc4, doc5]);
+  });
+
+  function assertQueryMatches(
+    query: Query,
+    matching: MutableDocument[],
+    nonMatching: MutableDocument[]
+  ): void {
+    for (const doc of matching) {
+      expect(queryMatches(query, doc)).true;
```
Comment (ehsannas) [inline packages/firestore/test/unit/core/query.test.ts:864]: should this be `.to.be.true` ?

File: packages/firestore/test/unit/core/query.test.ts [hunk 2]
```diff
@@ -800,6 +803,71 @@ describe('Query', () => {
...
+    }
+    for (const doc of nonMatching) {
+      expect(queryMatches(query, doc)).false;
```
Comment (ehsannas) [inline packages/firestore/test/unit/core/query.test.ts:867]: should this be `.to.be.false` ?

File: packages/firestore/test/unit/util/logic_utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+      andFilter(A, D),
+      andFilter(B, C),
+      andFilter(B, D)
+    );
+    expect(
+      filterEquals(
+        applyDistribution(orFilter(A, B), orFilter(C, D)),
+        expectedResult
+      )
+    );
```
Comment (ehsannas) [inline packages/firestore/test/unit/util/logic_utils.test.ts:181]: missing `.to.be.true`

**Review by ehsannas — COMMENTED** [review]
Great work. Just a few nits outlined below.

File: packages/firestore/src/core/query.ts [hunk 4]
```diff
@@ -468,7 +468,7 @@ function queryMatchesPathAndCollectionGroup(
  * in the results.
  */
 function queryMatchesOrderBy(query: Query, doc: Document): boolean {
-  for (const orderBy of query.explicitOrderBy) {
+  for (const orderBy of queryOrderBy(query)) {
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/src/core/query.ts:477]: Thanks for identifying this. Comment added.

File: packages/firestore/src/local/indexeddb_index_manager.ts [hunk 5]
```diff
@@ -544,7 +581,7 @@ export class IndexedDbIndexManager implements IndexManager {
 
     let boundIdx = 0;
     for (const segment of fieldIndexGetDirectionalSegments(fieldIndex)) {
-      const value = bound[boundIdx++];
+      const value = values[boundIdx++];
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/src/local/indexeddb_index_manager.ts]: Good call, updated.

File: packages/firestore/src/util/logic_utils.ts [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+function applyDistributionFieldAndCompositeFilters(
+  fieldFilter: FieldFilter,
+  compositeFilter: CompositeFilter
+): Filter {
+  // There are two cases:
+  // A & (B & C) --> (A & B & C)
+  // A & (B | C) --> (A & B) | (A & C)
+  if (compositeFilterIsConjunction(compositeFilter)) {
+    // Case 1
+    return CompositeFilter.create(
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/src/util/logic_utils.ts]: I updated this, but now I feel like it's inconsistent with the lines below (cases 2+), in Web and Android. However, it's not a big issue to me.

File: packages/firestore/src/util/logic_utils.ts [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+  // There are four cases:
+  // (A & B) & (C & D) --> (A & B & C & D)
+  // (A & B) & (C | D) --> (A & B & C) | (A & B & D)
+  // (A | B) & (C & D) --> (C & D & A) | (C & D & B)
+  // (A | B) & (C | D) --> (A & C) | (A & D) | (B & C) | (B & D)
+
+  // Case 1 is a merge.
+  if (compositeFilterIsConjunction(lhs) && compositeFilterIsConjunction(rhs)) {
+    return CompositeFilter.create(
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/src/util/logic_utils.ts]: I updated this, but now I feel like it's inconsistent with the lines below (cases 2+), in Web and Android. However, it's not a big issue to me.

File: packages/firestore/test/unit/core/query.test.ts (hunk 1 above)
Comment (MarkDuckworth) [reply packages/firestore/test/unit/core/query.test.ts]: Thank you. Fixed.

File: packages/firestore/test/unit/core/query.test.ts (hunk 2 above)
Comment (MarkDuckworth) [reply packages/firestore/test/unit/core/query.test.ts]: Thank you. Fixed.

File: packages/firestore/test/unit/util/logic_utils.test.ts (hunk 3 above)
Comment (MarkDuckworth) [reply packages/firestore/test/unit/util/logic_utils.test.ts]: Thank you. Fixed.

File: packages/firestore/src/core/target.ts [hunk 8]
```diff
@@ -786,6 +786,15 @@ export class CompositeFilter extends Filter {
   }
 }
 
+// TODO(orquery) move compositeFilterWithAddedFilters to filter.ts in future refactor
+export function compositeFilterWithAddedFilters(
```
Comment (ehsannas) [inline packages/firestore/src/core/target.ts:790]: This function shouldn't always perform `CompositeOperator.AND`. It should use the same operator as the given `compositeFilter` and add `otherFilters` to it.

Please add a comment for the function as well. ([Reference](https://cs.opensource.google/firebase-sdk/firebase-android-sdk/+/master:firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java;l=117-124?q=withadded&ss=firebase-sdk%2Ffirebase-android-sdk))

**Review by ehsannas — CHANGES_REQUESTED** [review]
One last thing

File: packages/firestore/src/core/target.ts (hunk 8 above)
Comment (MarkDuckworth) [reply packages/firestore/src/core/target.ts:794]: Fixed. TY.

**Review by ehsannas — APPROVED** [review]
Thanks!

**Pull request merged**